src/notebooks/controllers/controllerRegistration.ts

```diff
@@ -82,22 +94,261 @@ export class ControllerRegistration implements IControllerRegistration {
+        this.serverUriStorage.onDidChangeConnectionType(this.onDidChangeFilter, this, this.disposables);
+        this.serverUriStorage.onDidChangeUri(this.onDidChangeUri, this, this.disposables);
+        this.serverUriStorage.onDidRemoveUris(this.onDidRemoveUris, this, this.disposables);
+
+        this.onDidChange(
+            ({ added }) => {
+                added.forEach((controller) => {
+                    controller.onNotebookControllerSelectionChanged(
+                        (e) => {
+                            if (
+                                !e.selected &&
+                                this.isFiltered(controller.connection) &&
+                                this.canControllerBeDisposed(controller)
+                            ) {
+                                // This item was selected but is no longer allowed in the kernel list. Remove it
+                                traceWarning(
+                                    `Removing controller ${controller.id} for ${controller.connection.kind} from kernel list`
+                                );
+                                controller.dispose();
+                            }
+                        },
+                        this,
+                        this.disposables
+                    );
+                });
+            },
+            this,
+            this.disposables
+        );
+        // Sign up for document either opening or closing
+        this.notebook.onDidOpenNotebookDocument(this.onDidOpenNotebookDocument, this, this.disposables);
+        // If the extension activates after installing Jupyter extension, then ensure we load controllers right now.
+        this.notebook.notebookDocuments.forEach((notebook) => this.onDidOpenNotebookDocument(notebook).catch(noop));
+
+        this.loadControllers();
+        let previousKernelPickerType = this.featuresManager.features.kernelPickerType;
+        this.featuresManager.onDidChangeFeatures(
+            () => {
+                if (previousKernelPickerType === this.featuresManager.features.kernelPickerType) {
+                    return;
+                }
+                previousKernelPickerType = this.featuresManager.features.kernelPickerType;
+                // With the old kernel picker some controllers can get disposed.
+                // Hence to be on the safe side, when switching between the old and new kernel picker, reload the controllers.
+                this.loadControllers();
+            },
+            this,
+            this.disposables
+        );
+    }
+    private loadControllers() {
+        this.controllersPromise = this.loadControllersImpl();
+        sendKernelListTelemetry(this.registered.map((v) => v.connection));
+
+        traceInfoIfCI(`Providing notebook controllers with length ${this.registered.length}.`);
+    }
+    public get loaded() {
+        return this.controllersPromise;
+    }
+    private async onDidOpenNotebookDocument(document: NotebookDocument) {
```

Comment:
Brought all of this from another class.
Most of the code is very closely related, hence merged into a single class for easier maintenance
Besides a lot of this code will get deleted once we ship the new kernel picker.